validator/slashing-protection/attester_protect.go

```diff
@@ -0,0 +1,140 @@
+package slashingprotection
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/validator/db/kv"
+)
+
+func (s *Service) IsSlashableAttestation(
+	ctx context.Context,
+	indexedAtt *ethpb.IndexedAttestation,
+	pubKey [48]byte,
+	domain *ethpb.DomainResponse,
+) (bool, error) {
+	metricsKey := fmt.Sprintf("%#x", pubKey[:])
+	signingRoot, err := helpers.ComputeSigningRoot(indexedAtt.Data, domain.SignatureDomain)
+	if err != nil {
+		return false, err
+	}
+	s.attestingHistoryByPubKeyLock.Lock()
+	defer s.attestingHistoryByPubKeyLock.Unlock()
+	attesterHistory, ok := s.attesterHistoryByPubKey[pubKey]
+	if !ok {
+		return false, fmt.Errorf("could not get local slashing protection data for validator %#x", pubKey)
+	}
+	// Check if the attestation is slashable by local and remote slashing protection
+	if s.remoteProtector != nil {
+		slashable, err := s.remoteProtector.IsSlashableAttestation(ctx, indexedAtt, pubKey, domain)
+		if err != nil {
+			return false, err
+		}
+		remoteSlashableAttestationsTotal.WithLabelValues(metricsKey).Inc()
+		return slashable, nil
+	}
+	if isNewAttSlashable(
+		ctx,
+		attesterHistory,
+		indexedAtt.Data.Source.Epoch,
+		indexedAtt.Data.Target.Epoch,
+		signingRoot,
+	) {
+		localSlashableAttestationsTotal.WithLabelValues(metricsKey).Inc()
+		return true, nil
+	}
+	// We update the attester history with new values.
+	attesterHistory, err = attesterHistory.UpdateHistoryForAttestation(
+		ctx,
+		indexedAtt.Data.Source.Epoch,
+		indexedAtt.Data.Target.Epoch,
+		signingRoot,
+	)
+	if err != nil {
+		return false, errors.Wrap(err, "could not update attesting history data")
+	}
+	s.attesterHistoryByPubKey[pubKey] = attesterHistory
+	return false, nil
+}
+
+// isNewAttSlashable uses the attestation history to determine if an attestation of sourceEpoch
+// and targetEpoch would be slashable. It can detect double, surrounding, and surrounded votes.
+func isNewAttSlashable(ctx context.Context, history kv.EncHistoryData, sourceEpoch, targetEpoch uint64, signingRoot [32]byte) bool {
+	if history == nil {
+		return false
+	}
```

Comment:
This seems dangerous. Recommend: refactor to return (bool, error), return error if nil history provided.